src/input.rs

```diff
@@ -0,0 +1,48 @@
+use std::fs;
+use std::io::{self, Read, Seek, SeekFrom};
+
+pub enum Input<'a> {
+    File(fs::File),
+    Stdin(io::StdinLock<'a>),
+}
+
+impl<'a> Read for Input<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match *self {
+            Input::File(ref mut file) => file.read(buf),
+            Input::Stdin(ref mut stdin) => stdin.read(buf),
+        }
+    }
+}
+
+impl<'a> Seek for Input<'a> {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        match *self {
+            Input::File(ref mut file) => {
+                let seek_res = file.seek(pos);
+                if let Err(Some(29)) = seek_res.as_ref().map_err(|err| err.raw_os_error()) {
```

Comment:
can we use [`libc::ESPIPE`](https://docs.rs/libc/0.2.70/libc/constant.ESPIPE.html) here instead of 29?

Reply:
yeah, good idea